### PR TITLE
update to steno 0.2.0

### PR DESCRIPTION
**Pull request by davepacheco**
## Summary

This change updates Omicron to the in-progress Steno 0.2.0.  Steno 0.2 eliminates saga templates and lets consumers customize the saga graph each time the saga is run.  It also allows inserting subsagas into sagas.  These together enable things like #812.  Landing this change is blocked on oxidecomputer/steno#29.  That change is just about ready to land.

## Details

The upside of Steno 0.2 is the ability to run arbitrary saga graphs.  The cost is that the set of _actions_ is now fixed at build time and each action must be registered.  This created a small challenge in the instance-create saga where we use a `for` loop to stamp out N similar versions of the same action to do things like create N network interfaces.  Ironically, this was done because of the problem that Steno is now fixing.  I've made this work here, but there's room for cleaning this up further (see the comment in the source).

As part of this change, it was convenient to create a more first-class mechanism for registering the various sagas that Nexus runs.

In terms of the mechanical changes to each of the four sagas:

- The actions are defined with `lazy_static!` so that they can be easily referenced (with static type checking) from the code that builds both the DAG and the action registry. 
- There's now a unit struct that impls a `NexusSaga` trait that defines everything that's needed to stamp out a new saga, given parameters.  This comprises the saga's name, params type, a function that registers its actions, and a function that produces a new saga given some params.
- The saga params type is no longer part of the saga type in Steno 0.2, so all the calls to `saga_params()` are now fallible.
- As a result, all of the Nexus sagas have the same saga type.  So the action functions all accept the same `ActionContext`.

There were a few other small issues fixed here.

## Blockers / next steps

This change depends on oxidecomputer/steno#29.  My plan is:

- [x] get approval on this change
- [x] get approval on oxidecomputer/steno#29
- [x] land oxidecomputer/steno#29 (blocked on feedback from this change because I don't want to land that until we're happy with the consumer here in omicron)
- [x] update various Cargo.toml and Cargo.lock files in this PR to point to Steno branch "main" again
- [ ] land this change

### CODE REVIEW

**Comment by andrewjstone** [comment]
Woohoo!

File: nexus/types/src/external_api/params.rs
Comment (davepacheco) [inline nexus/types/src/external_api/params.rs:477]: tl;dr: I think you can ignore this, but here are the gory details.

Without this change, I saw lots of failures because we could not deserialize `params::InstanceCreate` from a `serde_json::Value`.  The error looked like this:

```
invalid type: string \"\", expected a borrowed string
```

This was reproducible with this test in params.rs:

```rust
#[test]
fn test_dap2() {
    let input = serde_json::json! {
        {   
          "name": "i1",
          "description": "asdf",
          "ncpus": 1,
          "memory": 1073741824,
          "hostname": "asdf",
          "user_data": "", 
          "network_interfaces": {
            "type": "default"
          },  
          "external_ips": [], 
          "disks": []
        }   
    };  
    let busted_as_string = serde_json::to_string_pretty(&input).unwrap();
    println!("busted input:\n{}", busted_as_string);
    println!("\nresult of parsing the serde_json::Value:");
    println!("{:?}", serde_json::from_value::<InstanceCreate>(input));
    println!("\nresult of parsing the same value, serialized to a string:");
    println!("{:?}", serde_json::from_str::<InstanceCreate>(&busted_as_string));
}
```

which produced:

```
busted input:
{
  "description": "asdf",
  "disks": [],
  "external_ips": [],
  "hostname": "asdf",
  "memory": 1073741824,
  "name": "i1",
  "ncpus": 1,
  "network_interfaces": {
    "type": "default"
  },
  "user_data": ""
}

result of parsing the serde_json::Value:
Err(Error("invalid type: string \"\", expected a borrowed string", line: 0, column: 0))

result of parsing the same value, serialized to a string:
Ok(InstanceCreate { identity: IdentityMetadataCreateParams { name: Name("i1"), description: "asdf" }, ncpus: InstanceCpuCount(1), memory: ByteCount(1073741824), hostname: "asdf", user_data: [], network_interfaces: Default, external_ips: [], disks: [] })
```

I'm not exactly sure what a `Deserialize` impl is _supposed_ to do here.  The [serde docs on Deserializer lifetimes](https://serde.rs/lifetimes.html) say:

> Not all deserializers support handing out borrowed data. For example when deserializing from an IO stream no data can be borrowed.

Given that, it seems safer to avoid deserializing borrowed values.

Why did I hit this now?  I think it's because with this change, the instance create saga stores `params::InstanceCreate` and deserializes them during the saga.  That causes them to go through a `serde_json::Value` where I suspect they didn't before.  However, the saga params _do_ go through a `serde_json::Value` at least during recovery so I suspect that saga recovery for the instance create saga was already broken.  I haven't tested that.

**Comment by davepacheco** [comment]
To be clear, this is ready for review.  I'm leaving it "draft" because it'll need to have the Cargo.toml/Cargo.lock references updated before landing.

File: nexus/db-model/src/saga_types.rs
Comment (andrewjstone) [inline nexus/db-model/src/saga_types.rs:206]: Nifty little safety feature :)

File: nexus/src/app/disk.rs
Comment (andrewjstone) [inline nexus/src/app/disk.rs:171]: Yay! Types!

File: nexus/src/app/instance.rs
Comment (andrewjstone) [inline nexus/src/app/instance.rs:177]: Nit: Was this newline on purpose?

File: nexus/src/app/sagas/disk_create.rs
Comment (andrewjstone) [inline nexus/src/app/sagas/disk_create.rs:54]: I like the name structure.

File: nexus/src/app/sagas/instance_create.rs
Comment (andrewjstone) [inline nexus/src/app/sagas/instance_create.rs:179]: s/for append/to append/ ?

File: nexus/src/app/sagas/instance_create.rs
Comment (andrewjstone) [inline nexus/src/app/sagas/instance_create.rs:214]: Grammar: `Here, `

File: nexus/src/app/sagas/instance_create.rs
Comment (andrewjstone) [inline nexus/src/app/sagas/instance_create.rs:748]: This seems important: Maybe open an issue for it?

File: nexus/src/app/sagas/instance_create.rs
Comment (andrewjstone) [inline nexus/src/app/sagas/instance_create.rs:879]: Issue?

File: nexus/src/app/sagas/instance_create.rs
Comment (andrewjstone) [inline nexus/src/app/sagas/instance_create.rs:222]: Do we want to fix this here, or leave it to a follow up PR?

**Review by andrewjstone — APPROVED** [review]
Looks great @davepacheco. Overall, a straightforward mechanical transformation. I really like how we ended up registering actions in particular.

File: nexus/src/app/sagas/instance_create.rs
Comment (davepacheco) [reply nexus/src/app/sagas/instance_create.rs:222]: I considered fixing this here but the surgery got more complicated than I wanted to take on here.  (I think you want to split the create-nic action into two separate actions: one for the default config and one for custom.  Both would have the same undo function.)

Comment (andrewjstone) [reply nexus/src/app/sagas/instance_create.rs:222]: Ok, that makes sense. I think this works as a first cut that enables the same behavior with steno 0.2.

File: nexus/src/app/sagas/instance_create.rs
Comment (davepacheco) [reply nexus/src/app/sagas/instance_create.rs]: Fixed in 33951e28e39cfd0dfeacf5066700dff6f206b045.

File: nexus/src/app/sagas/instance_create.rs
Comment (davepacheco) [reply nexus/src/app/sagas/instance_create.rs]: Fixed in 33951e28e39cfd0dfeacf5066700dff6f206b045.

File: nexus/src/app/instance.rs
Comment (davepacheco) [reply nexus/src/app/instance.rs]: I don't remember.  I just removed it in 33951e28e39cfd0dfeacf5066700dff6f206b045.

File: nexus/db-model/src/saga_types.rs
Comment (davepacheco) [reply nexus/db-model/src/saga_types.rs:206]: I think I picked this up from @cbiffle.  I like it.

File: nexus/src/app/sagas/instance_create.rs
Comment (davepacheco) [reply nexus/src/app/sagas/instance_create.rs:748]: Filed https://github.com/oxidecomputer/omicron/issues/1536 and updated this comment to reference it.

File: nexus/src/app/sagas/instance_create.rs
Comment (davepacheco) [reply nexus/src/app/sagas/instance_create.rs:878]: Filed https://github.com/oxidecomputer/omicron/issues/1536 and updated this comment to reference it.

File: nexus/src/app/saga.rs
Comment (bnaecker) [inline nexus/src/app/saga.rs:69]: Nit: The docstring still refers to a "saga template", but IIUC templates are a thing of the past.

File: nexus/src/app/sagas/instance_create.rs
Comment (bnaecker) [inline nexus/src/app/sagas/instance_create.rs:380]: The comment here, deleting by index, is no longer accurate, I think. It's done by primary key / ID.

File: nexus/src/app/sagas/instance_create.rs
Comment (bnaecker) [inline nexus/src/app/sagas/instance_create.rs:431]: My original error, but spelling: `interfaces` -> `interface`.

**Review by bnaecker — APPROVED** [review]
Looks great, the Steno updates are really nice! I've a few nits around comments, but overall LGTM.

File: nexus/src/app/saga.rs
Comment (davepacheco) [reply nexus/src/app/saga.rs]: Good call.  Fixed in 96dba6366f6cb8b47ff10a3a4b0bf0277f9d060e.  I also searched for other references and fixed them.

File: nexus/src/app/sagas/instance_create.rs
Comment (davepacheco) [reply nexus/src/app/sagas/instance_create.rs]: Nice catch.  Fixed in 96dba6366f6cb8b47ff10a3a4b0bf0277f9d060e.

File: nexus/src/app/sagas/instance_create.rs
Comment (davepacheco) [reply nexus/src/app/sagas/instance_create.rs]: Fixed in 96dba6366f6cb8b47ff10a3a4b0bf0277f9d060e.

**Comment by davepacheco** [comment]
With this change approved, oxidecomputer/steno#29 has landed.  I've updated this change to point steno back to "main" and this is ready to go too.  Thanks!

**Comment by davepacheco** [comment]
This is blocked on the build problem that I haven't yet figured out how to resolve.  Summary (discussed in oxide-rust chat):

- The root of the omicron repo is a workspace containing various crates.  Among those crates is omicron-common.
- Some of these crates depend on each other.  They generally use Cargo "path" dependencies.
- omicron-common depends on "steno" using a git dependency on branch "main".  At all times, the repo lockfile specifies a particular commit from that branch.
- A recent update to steno (which I'm trying to pull in with this PR) is a breaking change.  The version has bumped accordingly from 0.1.0 to 0.2.0. (The Cargo docs say these are considered incompatible versions, which is good for our purposes.)
- A different top-level crate in the omicron workspace is omicron-sled-agent.  It depends on a crate "crucible" using a git dependency with a specific git revision.  This package [depends on omicron-common](https://github.com/oxidecomputer/crucible/blob/0c1d7245445dcc6b61fa18a4c7806080bf7853b5/upstairs/Cargo.toml#L25) via a git dependency on branch "main".
- This means that omicron's Cargo.lock contains two copies of omicron-common and some related packages: one from the local repo and one from git.  This [is not ideal](https://github.com/oxidecomputer/omicron/issues/1537) but does (mostly) work.
- Right now, omicron branch "main" refers to steno 0.1.0 (pre-breaking-change).  A top level `cargo test` works fine.

Despite #1537, it seems like this could work correctly if Cargo.lock said:

- the in-repo omicron-common depends on steno 0.2.0 from git sha $latest
- the git-source omicron-common depends on steno 0.1.0 from git sha $older

That's not what Cargo.lock says, though.  In the PR branch, all references in Cargo.lock refer to steno 0.2.0 from git sha $latest.  (I don't have the full history here but I suspect I was overzealous in running `cargo update -p steno` while I was working on this PR because I didn't know there was a tree of dependencies that I'd not want to update.)  The net result is that we build both the in-repo omicron-common and the git-source omicron-common against steno 0.2.0, and the git-source one doesn't build because of course it's from "main" and hasn't been updated yet.

**Comment by davepacheco** [comment]
Okay, I think I've resolved the build problem but I'd love a +1 on my solution.  There's a bunch of noise so I recommend looking at [this entire range](https://github.com/oxidecomputer/omicron/pull/1532/files/96dba6366f6cb8b47ff10a3a4b0bf0277f9d060e..67e3e3088877831d404918b19ad6e7749d27584a) or maybe [the Cargo.toml/Cargo.lock changes from "main"](https://github.com/oxidecomputer/omicron/pull/1532/files?manifests=true&show-viewed-files=true&file-filters%5B%5D=).  The most interesting bit is the patch section I added to the workspace Cargo.toml:

https://github.com/oxidecomputer/omicron/pull/1532/files?manifests=true&show-viewed-files=true&file-filters%5B%5D=#diff-2e9d962a08321605940b5a657135052fbcef87b5e360662bb527c96d9a615542R92-R112

The comment should explain the reasoning.  As I understand it, once this change lands on "main", we can immediately remove it.

This problem would also be resolved by #1537.  More generally, it seems really fraught to have multiple repos that contain crates with circular dependencies of any kind (assuming the intent is that you can `cargo build` the whole workspace -- I realize that's not true for all repos, like hubris).

**Comment by andrewjstone** [comment]
> Okay, I think I've resolved the build problem but I'd love a +1 on my solution. There's a bunch of noise so I recommend looking at [this entire range](https://github.com/oxidecomputer/omicron/pull/1532/files/96dba6366f6cb8b47ff10a3a4b0bf0277f9d060e..67e3e3088877831d404918b19ad6e7749d27584a) or maybe [the Cargo.toml/Cargo.lock changes from "main"](https://github.com/oxidecomputer/omicron/pull/1532/files?manifests=true&show-viewed-files=true&file-filters%5B%5D=). The most interesting bit is the patch section I added to the workspace Cargo.toml:
> 
> https://github.com/oxidecomputer/omicron/pull/1532/files?manifests=true&show-viewed-files=true&file-filters%5B%5D=#diff-2e9d962a08321605940b5a657135052fbcef87b5e360662bb527c96d9a615542R92-R112
> 
> The comment should explain the reasoning. As I understand it, once this change lands on "main", we can immediately remove it.
> 
> This problem would also be resolved by #1537. More generally, it seems really fraught to have multiple repos that contain crates with circular dependencies of any kind (assuming the intent is that you can `cargo build` the whole workspace -- I realize that's not true for all repos, like hubris).

Huge +1 to this. Let's get this code in and then revert the patch as you propose. And yeah, I'm a firm believer in the solution in #1537 and the strategy there in general.

Thanks again for doing this migration and all the subsequent shepherding @davepacheco. This is exciting!

**Pull request merged**
